src/zenml/config/compiler.py

```diff
@@ -648,6 +653,103 @@ def _compute_pipeline_spec(
             "parameters": pipeline._parameters,
         }
 
+        # Store a JSON schema of the pipeline input parameters for CLI/UI
+        # tooling.
+        if pipeline_parameters_model := pipeline.get_parameters_model():
+            additional_spec_args["parameters_schema"] = (
+                pipeline_parameters_model.model_json_schema()
+            )
+
+        # Best-effort: build a response schema snapshot for step outputs (tooling).
+        try:
+            # Map invocation id -> StepSpec for quick lookup
+            name_to_spec: Dict[str, StepSpec] = {
+                s.pipeline_parameter_name: s for s in step_specs
+            }
+            all_names: set[str] = set(name_to_spec.keys())
+            # Build downstream map from upstream relationships
+            downstream: Dict[str, Set[str]] = {n: set() for n in all_names}
+            for s in step_specs:
+                for up in s.upstream_steps:
+                    if up in downstream:
+                        downstream[up].add(s.pipeline_parameter_name)
+            # NOTE: The serving response uses flat keys in the form
```

Comment:
This is the compiler for a pipeline, it certainly doesn't know what serving even is. And it for sure doesn't care about the structure of outputs in a served deployment. It should store the output in a format that is generic enough so that everyone who decides to use it in the future can modify it in a way that works for their use case.